src/main/Main.hs

```diff
@@ -470,21 +473,41 @@ setupCmd SetupCmdOpts{..} go@GlobalOpts{..} = do
               )
           Nothing
 
-withConfig :: GlobalOpts
+-- | Enforce mutual exclusion of every action running via this function
+-- on this users account.  Currently, stack uses this to completely
+-- exclude concurrent execution of stack commands.  In the future,
+-- stack may refine this to a finer-grained locking approach.
+withUserFileLock :: Config
+                 -> (FileLock -> IO a)
+                 -> IO a
+withUserFileLock cfg act = do
+    lockfile <- parseRelFile "lockfile"
```

Comment:
Ah, ok, these uses of TH are neat but new to me.  I'll try this.

Btw, it seems like many of them could probably use the new typed TH interface, no?   (The one confusing thing about reading the APIs that use TH is the loss of type info in all the `Q Exp`s.)